File: ObjcExample/ObjcExample/Demo/GKPageScrollView/HeaderScroll/GKHeaderScrollViewController.m
//
//  GKHeaderScrollViewController.m
//  GKPageScrollViewObjc
//
//  Created by QuintGao on 2019/5/31.
//  Copyright © 2019 QuintGao. All rights reserved.
//

#import "GKHeaderScrollViewController.h"
#import "GKHeaderScrollView.h"
#import "GKPageScrollView.h"
#import <MJRefresh/MJRefresh.h>
#import "GKBaseListViewController.h"

@interface GKHeaderScrollViewController()<GKPageScrollViewDelegate, GKPageTableViewGestureDelegate>

@property (nonatomic, strong) GKPageScrollView      *pageScrollView;

@property (nonatomic, strong) GKHeaderScrollView    *headerView;

@property (nonatomic, strong) UIView                *pageView;
@property (nonatomic, strong) JXCategoryTitleView   *categoryView;
@property (nonatomic, strong) UIScrollView          *scrollView;

@property (nonatomic, strong) NSArray               *titles;
@property (nonatomic, strong) NSMutableArray        *childVCs;

@end

@implementation GKHeaderScrollViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.gk_navTitleColor = [UIColor blackColor];
    self.gk_navTitleFont = [UIFont boldSystemFontOfSize:18.0f];
    self.gk_navBackgroundColor = [UIColor whiteColor];
    self.gk_statusBarStyle = UIStatusBarStyleDefault;
    self.gk_backStyle = GKNavigationBarBackStyleBlack;
    self.gk_navTitle = @"headerScroll";
    
    [self.view addSubview:self.pageScrollView];
    [self.pageScrollView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self.view).insets(UIEdgeInsetsMake(GK_STATUSBAR_NAVBAR_HEIGHT, 0, 0, 0));
    }];
    
    self.categoryView.titles = self.titles;
    [self.categoryView reloadData];
    
    // 根据标题创建控制器并添加到scrollView
    [self.titles enumerateObjectsUsingBlock:^(id  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        GKBaseListViewController *vc = [GKBaseListViewController new];
        
        [self.childVCs addObject:vc];
        
        [self.scrollView addSubview:vc.view];
        vc.view.frame = CGRectMake(idx * self.view.frame.size.width, 0, self.view.frame.size.width, self.view.frame.size.height - kBaseSegmentHeight - kNavBarHeight);
    }];
    self.scrollView.contentSize = CGSizeMake(self.titles.count * self.view.frame.size.width, 0);
    
    // 刷新
    [self.pageScrollView reloadData];
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    
    CGRect frame = self.headerView.frame;
    frame.size.width = self.view.frame.size.width;
    self.headerView.frame = frame;
    
    frame = self.categoryView.frame;
    if (frame.size.width != self.view.frame.size.width) {
        frame.size.width = self.view.frame.size.width;
        self.categoryView.frame = frame;
        [self.categoryView reloadData];
    }
    
    [self.childVCs enumerateObjectsUsingBlock:^(UIViewController *obj, NSUInteger idx, BOOL * _Nonnull stop) {
        obj.view.frame = CGRectMake(idx * self.view.frame.size.width, 0, self.view.frame.size.width, self.view.frame.size.height);
    }];
    self.scrollView.contentSize = CGSizeMake(self.childVCs.count * self.view.frame.size.width, 0);
}

#pragma mark - GKPageScrollViewDelegate
- (UIView *)headerViewInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.headerView;
}

- (UIView *)pageViewInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.pageView;
}

- (NSArray<id<GKPageListViewDelegate>> *)listViewsInPageScrollView:(GKPageScrollView *)pageScrollView {
    return self.childVCs;
}

#pragma mark - GKPageTableViewGestureDelegate
- (BOOL)pageTableView:(GKPageTableView *)tableView gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldRecognizeSimultaneouslyWithGestureRecognizer:(UIGestureRecognizer *)otherGestureRecognizer {
    // 禁止UIScrollView左右滑动时，上下左右都可以滑动
    UIScrollView *scrollView = [self.headerView valueForKey:@"collectionView"];
    
    if (otherGestureRecognizer == scrollView.panGestureRecognizer) {
        return NO;
    }
    
    if (otherGestureRecognizer == self.scrollView.panGestureRecognizer) {
        return NO;
    }
    
    return [gestureRecognizer.view isKindOfClass:[UIScrollView class]] && [otherGestureRecognizer.view isKindOfClass:[UIScrollView class]];
}

#pragma mark - 懒加载
- (GKPageScrollView *)pageScrollView {
    if (!_pageScrollView) {
        _pageScrollView = [[GKPageScrollView alloc] initWithDelegate:self];
        _pageScrollView.ceilPointHeight = 0;
        _pageScrollView.mainTableView.gestureDelegate = self;
    }
    return _pageScrollView;
}

- (GKHeaderScrollView *)headerView {
    if (!_headerView) {
        CGFloat headerH = (self.view.frame.size.width - 40) / 4 + 20;
        
        _headerView = [[GKHeaderScrollView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, headerH)];
    }
    return _headerView;
}

- (NSArray *)titles {
    if (!_titles) {
        _titles = @[@"动态", @"文章", @"更多"];
    }
    return _titles;
}

- (UIView *)pageView {
    if (!_pageView) {
        _pageView = [UIView new];
        
        [_pageView addSubview:self.categoryView];
        [_pageView addSubview:self.scrollView];
        
        [self.categoryView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.right.top.equalTo(self->_pageView);
            make.height.mas_equalTo(kBaseSegmentHeight);
        }];
        
        [self.scrollView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.right.bottom.equalTo(self->_pageView);
            make.top.equalTo(self.categoryView.mas_bottom);
        }];
    }
    return _pageView;
}

- (JXCategoryTitleView *)categoryView {
    if (!_categoryView) {
        _categoryView = [[JXCategoryTitleView alloc] init];
        _categoryView.titleFont = [UIFont systemFontOfSize:15.0f];
        _categoryView.titleSelectedFont = [UIFont systemFontOfSize:15.0f];
        _categoryView.titleColor = [UIColor grayColor];
        _categoryView.titleSelectedColor = [UIColor redColor];
        
        JXCategoryIndicatorLineView *lineView = [JXCategoryIndicatorLineView new];
        lineView.lineStyle = JXCategoryIndicatorLineStyle_Normal;
        lineView.indicatorHeight = ADAPTATIONRATIO * 4.0f;
        lineView.verticalMargin = ADAPTATIONRATIO * 2.0f;
        _categoryView.indicators = @[lineView];
        
        // 设置关联的scrollView
        _categoryView.contentScrollView = self.scrollView;
        
        UIView  *btmLineView = [UIView new];
        btmLineView.backgroundColor = GKColorRGB(110, 110, 110);
        [_categoryView addSubview:btmLineView];
        [btmLineView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.left.right.bottom.equalTo(self->_categoryView);
            make.height.mas_equalTo(ADAPTATIONRATIO * 2.0f);
        }];
    }
    return _categoryView;
}

- (UIScrollView *)scrollView {
    if (!_scrollView) {
        _scrollView = [[UIScrollView alloc] init];
        _scrollView.pagingEnabled = YES;
        _scrollView.bounces = NO;
        _scrollView.gk_openGestureHandle = YES;
    }
    return _scrollView;
}

- (NSMutableArray *)childVCs {
    if (!_childVCs) {
        _childVCs = [NSMutableArray new];
    }
    return _childVCs;
}

@end

File: SwiftExample/SwiftExample/Demo/GKPageScrollView/HeaderScroll/GKHeaderScrollViewController.swift
//
//  GKHeaderScrollViewController.swift
//  GKPageScrollViewSwift
//
//  Created by gaokun on 2019/6/3.
//  Copyright © 2019 gaokun. All rights reserved.
//

import UIKit
import JXSegmentedView
import MJRefresh
import GKPageScrollView

class GKHeaderScrollViewController: GKDemoBaseViewController {
    
    var titleDataSource = JXSegmentedTitleDataSource()
    
    lazy var pageScrollView: GKPageScrollView! = {
        let pageScrollView = GKPageScrollView(delegate: self)
        pageScrollView.ceilPointHeight = 0
        return pageScrollView
    }()
    
    lazy var headerView: GKHeaderScrollView = {
        var headerH = (kScreenW - 40)/4 + 20;
        
        let headerView = GKHeaderScrollView(frame: CGRect(x: 0, y: 0, width: kScreenW, height: headerH))
        headerView.delegate = self
        
        return headerView
    }()
    
    let titles = ["动态", "文章", "更多"]
    
    lazy var pageView: UIView = {
        let pageView = UIView()
        pageView.addSubview(self.segmentedView);
        pageView.addSubview(self.scrollView)
        
        return pageView
    }()
    
    lazy var segmentedView: JXSegmentedView = {
        titleDataSource.titleNormalColor = UIColor.gray
        titleDataSource.titleSelectedColor = UIColor.red
        titleDataSource.titleNormalFont = UIFont.systemFont(ofSize: 15.0)
        titleDataSource.titleSelectedFont = UIFont.systemFont(ofSize: 15.0)
        
        var segmentedView = JXSegmentedView(frame: CGRect(x: 0, y: 0, width: kScreenW, height: kBaseSegmentHeight))
        segmentedView.delegate = self
        segmentedView.dataSource = titleDataSource
        
        let lineView = JXSegmentedIndicatorLineView()
        lineView.lineStyle = .normal
        lineView.indicatorHeight = ADAPTATIONRATIO * 4.0
        lineView.verticalOffset = ADAPTATIONRATIO * 2.0
        segmentedView.indicators = [lineView]
        
        segmentedView.contentScrollView = self.scrollView
        
        let btmLineView = UIView()
        btmLineView.backgroundColor = UIColor.grayColor(g: 110)
        segmentedView.addSubview(btmLineView)
        btmLineView.snp.makeConstraints({ (make) in
            make.left.right.bottom.equalTo(segmentedView)
            make.height.equalTo(ADAPTATIONRATIO * 2.0)
        })
        
        return segmentedView
    }()
    
    lazy var scrollView: UIScrollView = {
        let scrollView = UIScrollView(frame: CGRect(x: 0, y: kBaseSegmentHeight, width: kScreenW, height: kScreenH - kBaseSegmentHeight - kNavBar_Height))
        scrollView.isPagingEnabled = true
        scrollView.bounces = false
        scrollView.gk_openGestureHandle = true
        
        return scrollView
    }()
    
    var childVCs = [GKBaseListViewController]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.gk_navTitle = "headerScroll";
        self.gk_navTitleColor = .black
        self.gk_navTitleFont = UIFont.boldSystemFont(ofSize: 18);
        self.gk_navBackgroundColor = .white
        self.gk_statusBarStyle = .default;
        self.gk_backStyle = .black;
        self.gk_navLineHidden = true
        
        self.view.addSubview(pageScrollView)
        
        pageScrollView.snp.makeConstraints { (make) in
            make.edges.equalTo(self.view).inset(UIEdgeInsets(top: kNavBar_Height, left: 0, bottom: 0, right: 0))
        }
        
        // 设置数据
        self.titleDataSource.titles = self.titles;
        self.titleDataSource.reloadData(selectedIndex: 0)
        
        // 根据标题创建控制器并添加到scrollView
        for i in 0..<self.titles.count {
            let vc = GKBaseListViewController(listType: .UITableView)
            self.childVCs.append(vc)
            
            self.scrollView.addSubview(vc.view)
            vc.view.frame = CGRect(x: CGFloat(i) * kScreenW, y: 0, width: kScreenW, height: kScreenH - kBaseSegmentHeight - kNavBar_Height)
        }
        self.scrollView.contentSize = CGSize(width: CGFloat(self.titles.count) * kScreenW, height: 0)
        
        // 刷新
        self.pageScrollView.reloadData()
    }
}

extension GKHeaderScrollViewController: GKPageScrollViewDelegate {
    func headerView(in pageScrollView: GKPageScrollView) -> UIView {
        return self.headerView
    }
    
    func pageView(in pageScrollView: GKPageScrollView) -> UIView {
        return self.pageView
    }
    
    func listView(in pageScrollView: GKPageScrollView) -> [GKPageListViewDelegate] {
        return self.childVCs
    }
}

extension GKHeaderScrollViewController: JXSegmentedViewDelegate {
    
}

extension GKHeaderScrollViewController: GKHeaderScrollViewDelegate {
    func headerScrollViewWillBeginScroll() {
        self.pageScrollView.horizonScrollViewWillBeginScroll()
    }
    
    func headerScrollViewDidEndScroll() {
        self.pageScrollView.horizonScrollViewDidEndedScroll()
    }
}
